SNLS_linalg.h

```diff
@@ -0,0 +1,339 @@
+/// prod has dimensions ldim x mdim
+/// This function will either accumulate the values of
+/// the multiplication on the product,
+/// or it will zero out the product ahead of time
+/// depedening on the run time flag.
+template <int ldim, int ndim, int mdim, bool zero_out>
+__snls_hdev__
+inline void matMultT(const double* const mat1,
+                     const double* const mat2,
+                     double* const prod)
+{
+    if (zero_out)
+    {
+        for (int iLM = 0; iLM < (ldim * mdim); iLM++) { prod[iLM] = 0.0; }
+    }
+    //prod_ik = mat1_ij * mat2_kj
+    for (int iL = 0; iL < ldim; iL++)
+    {
+        for (int jN = 0; jN < ndim; jN++)
+        {
+            for (int kM = 0; kM < mdim; kM++)
+            {
+                prod[iL * mdim + kM] += mat1[iL * ndim + jN] * mat2[kM * ndim + jN];
+            }
+        }
+    }
+}
+
+/// Performs the triple product operation
+/// needed to rotate a matrix of ndim x ndim
+/// by a rotatation matrix
+/// The product matrix is zeroed out in this operation
+/// Transpose does the operation:
+/// prod_il = rot_ji mat_jk rot_kl
+/// prod = rot^t * mat * rot
+/// Non-transpose operation does:
+/// prod_il = rot_ij mat_jk rot_lk
+/// prod = rot * mat * rot^T
+template<int ndim, bool transpose>
+__snls_hdev__
+inline void rotMatrix(const double* const mat,
+                    const double* const rot,
+                    double* const prod)
+{
+    // zero things out first
+    for (int iNN = 0; iNN < ndim * ndim; iNN++)
+    {
+        prod[iNN] = 0.0;
+    }
+    // Now for matrix multiplication
+    for (int iN = 0; iN < ndim; iN++)
+    {
+        for (int jN = 0; jN < ndim; jN++)
+        {
+            for (int kN = 0; kN < ndim; kN++)
+            {
+                for (int lN = 0; lN < ndim; lN++)
+                {
+                    // This is rot_ji mat_jk rot_kl
+                    if (transpose)
+                    {
+                    prod[iN * ndim + lN] += rot[jN * ndim + iN] *
+                                            mat[jN * ndim + kN] *
+                                            rot[kN * ndim + lN];
+                    }
+                    // This is rot_ij mat_jk rot_lk
+                    else
+                    {
+                    prod[iN * ndim + lN] += rot[iN * ndim + jN] *
+                                            mat[jN * ndim + kN] *
+                                            rot[lN * ndim + kN];
+                    }
+                }
+            }
+        }
+    }
+}
+
+template<int n>
+inline void
+printVec(const double* const y) {
+    for (int iX = 0; iX<n; ++iX) {
+        std::cout << std::setw(21) << std::setprecision(14) << y[iX] << " ";
+    }
+
+    std::cout << std::endl;
+}
+
+template<int n>
+inline void
+printMat(const double* const A) {
+    for (int iX = 0; iX<n; ++iX) {
+        for (int jX = 0; jX<n; ++jX) {
+            std::cout << std::setw(21) << std::setprecision(14) << A[SNLS_NN_INDX(iX, jX, n)] << " ";
```

Comment:
I do not see where SNLS_NN_INDX is coming from. It does not look like it is in SNLS_base.h

It is the same as SNLSTRDLDG_J_INDX? If so, why not combine them so that the same macro is used for both?

Reply:
It turns out I forgot to upload the changes that include this value in `SNLS_base.h`. I'll add that back in as part of the PR refactor as well as some modifications to the Trust-Region Dogleg solver to make use of some of the linalg functions.